_posts/2025-01-08-working-with-graph-databases-in-python.md

```diff
@@ -0,0 +1,278 @@
+                                                                       rdflib.term.Literal('rhyolite', lang='fr'),
+                                                                       rdflib.term.Literal('riolite', lang='it')]}
+```
+
+If we want to see its parents, we can access the object behind the predicate `skos:broader`.
+Here we will also use the predicate `skos:prefLabel` to access the English name for the parents.
+
+```python
+broader_uris = graph_dict[rhyolite_uri][skos_broader_uri]
+parents = []
+for parent_uri in broader_uris:
+    parent_labels = graph_dict[parent_uri][skos_pref_label_uri]
+    # Get just the English label for each parent
+    english_label = [label.toPython() for label in parent_labels if label.language == "en"][0]
+    parents.append(english_label)
+
+pprint(parents)
+```
+
+```
+['rhyolitoid']
+```
+
+Finally, in this format of linked data, recursive functions are incredibly useful to traverse a graph
+from one node to another. Or, in the case of `triples`, from one subject to another. Here, we will
+use recursion to find the list of all parent lithology classifications above `Rhyolite`.
+
+```python
+def find_lithology_parents(
+    graph_dict,
+    target_lithology_uri,
+    parent_labels = set(),
+):
+    """
+    Find all of the parent lithology classifications above the given target_lithology_uri.
+    A set of all parent labels is returned.
+    We use a set because this ensures only unique values are stored within it.
+
+    The argument 'parent_labels' can be ignored when calling this function,
+    as it is used to start an empty set which is populated as the recursion iterates.
+    """
+    # If the target lithology has parents
+    if skos_broader_uri in graph_dict[target_lithology_uri]:
+
+        for parent_uri in graph_dict[target_lithology_uri][skos_broader_uri]:
+            # Get the current parent English label
+            current_parent_labels = graph_dict[parent_uri][skos_pref_label_uri]
+            english_label = [label.toPython() for label in current_parent_labels if label.language == "en"][0]
+
+            # Add the current parent to the set of all parents
+            parent_labels.add(english_label)
+
+            # Find the parents of the parent
+            parent_parents = find_lithology_parents(graph_dict, parent_uri, parent_labels)
+            # Add the parents of the parent to the set of all parents
+            parent_labels = parent_labels.union(parent_parents)
+
+    return parent_labels
+
+
+rhyolite_parents = find_lithology_parents(graph_dict, rhyolite_uri)
+pprint(rhyolite_parents)
+```
+
+```
+{'acidic_igneous_material',
+ 'acidic_igneous_rock',
+ 'compound_material',
+ 'fine_grained_igneous_rock',
+ 'igneous_material',
+ 'igneous_rock',
+ 'rhyolitoid',
+ 'rock'}
+```
+
+## Alternative Solutions
+
+It is worth noting that there are other ways of parsing data within a graph database.
```

Comment:
It is worth noting that there are other ways of extracting data from a triplestore.